client/chat/create-channel.tsx

```diff
@@ -0,0 +1,105 @@
+import React, { useState } from 'react'
+import styled from 'styled-components'
+import { ChatServiceErrorCode } from '../../common/chat'
+import { CHANNEL_MAXLENGTH, CHANNEL_PATTERN } from '../../common/constants'
+import { useForm } from '../forms/form-hook'
+import SubmitOnEnter from '../forms/submit-on-enter'
+import { composeValidators, maxLength, regex, required } from '../forms/validators'
+import logger from '../logging/logger'
+import { useAutoFocusRef } from '../material/auto-focus'
+import { RaisedButton } from '../material/button'
+import { TextField } from '../material/text-field'
+import { isFetchError } from '../network/fetch-errors'
+import { useAppDispatch } from '../redux-hooks'
+import { useStableCallback } from '../state-hooks'
+import { colorError } from '../styles/colors'
+import { headline4, subtitle1 } from '../styles/typography'
+import { createChannel, navigateToChannel } from './action-creators'
+
+const CreateChannelRoot = styled.div`
+  display: flex;
+  flex-direction: column;
+  gap: 16px;
+  max-width: 400px;
+  padding: 16px 24px;
+`
+
+const Title = styled.div`
+  ${headline4};
+`
+
+const ErrorText = styled.div`
+  ${subtitle1};
+  color: ${colorError};
+`
+
+const channelValidator = composeValidators(
+  required('Enter a channel name'),
+  maxLength(CHANNEL_MAXLENGTH, `Enter at most ${CHANNEL_MAXLENGTH} characters`),
+  regex(CHANNEL_PATTERN, 'Channel name contains invalid characters'),
+)
+
+interface JoinChannelModel {
+  channel: string
+}
+
+export function CreateChannel() {
+  const dispatch = useAppDispatch()
+  const [errorMessage, setErrorMessage] = useState<string>()
+  const autoFocusRef = useAutoFocusRef<HTMLInputElement>()
+
+  const onFormSubmit = useStableCallback((model: JoinChannelModel) => {
+    const channelName = model.channel
+
+    dispatch(
+      createChannel(channelName, {
+        onSuccess: channel => navigateToChannel(channel.channelInfo.id, channel.channelInfo.name),
+        onError: err => {
+          let message = `An error occurred while joining ${channelName}`
+
+          if (isFetchError(err) && err.code) {
+            if (err.code === ChatServiceErrorCode.UserBanned) {
+              message = `You are banned from ${channelName}`
+            } else {
+              logger.error(`Unhandled code when joining ${channelName}: ${err.code}`)
+            }
+          } else {
+            logger.error(`Error when joining ${channelName}: ${err.stack ?? err}`)
+          }
+          setErrorMessage(message)
```

Comment:
I would suggest never doing this. The nice part about handling the error here is that we can have access to the actual error when rendering. If you turn everything into a single string and use that as state you have destroyed most of this advantage. Just save the error in state and do the logic to convert it to something visual in render (where you can like, do something nicer than just display a single line of text)

Reply:
Yeah, that's what I usually do too. The reason I moved it inside the error handler here is because I could read the `channelName` here that the form was submitted with, without saving it somewhere in state. I tried using the `getInputValue` first to read it directly from the form model, but then the error message gets updated as well if the search input changes.

Maybe there's something in the form hook that can be used to read the values that the form was last submitted with?

Reply:
Just set it in state? I don't see why you're trying to take some worse route just to avoid 2 lines of code

Reply:
Alternatively just don't include the channel name in the message? I don't really see how that serves much purpose in a real-world scenario